### ISSUE TITLE
[@types/react] RefObject.current should no longer be readonly

### ISSUE THREAD

**Issue by bschlenk** [DefinitelyTyped/DefinitelyTyped#31065]
It is now okay to assign to `ref.current`, see example: https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables

Trying to assign a value to it gives the error: `Cannot assign to 'current' because it is a constant or a read-only property.`

- [x] I tried using the `@types/react` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @johnnyreilly @bbenezech @pzavolinsky @digiguru @ericanderson @morcerf @tkrotoff @DovydasNavickas @onigoetz @theruther4d @guilhermehubner @ferdaber @jrakotoharisoa @pascaloliv @Hotell @franklixuefei

**Comment by ferdaber** [comment]
A PR would be appreciated! It should be a very easy one line change 😊

**Comment by jagatstha** [comment]
@ferdaber, 
Hi, I like to pick this issue. As I am new to typescript (just a week) and it will be my first contribution to open source. 

I have gone through node_modules/@types/react/index.d.ts and there is the following code at line no 61 which declare current as readonly.

```
interface RefObject<T> {
        readonly current: T | null;
    }
```
if this is the issue, then I can solve. can you guide for my first PR?
 Thank you for your time :)

**Comment by ferdaber** [comment]
Yes, it's just removing the `readonly` modifier on that line.

**Comment by pshrmn** [comment]
`React.useRef` returns a `MutableRefObject` whose `current` property is not `readonly`. I suppose the question is whether the ref object types should be unified.

**Comment by ferdaber** [comment]
They should be unified, the React runtime has no limitation on the `current` property created by `React.createRef()`, the object itself is sealed but not frozen.

**Comment by ferdaber** [comment]
That's fair, I've been using `React.createRef()` as a helper function to just create a pointer, as React won't manage it unless it's passed in as a `ref` prop, but you can just as well create a simple pointer object with the same structure.

**Comment by Kovensky** [comment]
We could modify `createRef` to have the same overload. If you include `| null` it'd return `MutableRefObject`, if you don't it'd be (immutable) `RefObject`.

**Comment by Kovensky** [comment]
🤔

Should we really make it easier to write incorrect code for those that _do_ have `strictNullTypes` enabled?

**Comment by bschlenk** [comment]
When I created this issue I didn't realize there already was this `| null` overload. It isn't a very common pattern so I didn't expect something like that to exist. I don't know how I missed it in the documentation though, it is very well documented and explained. I'm fine with closing this as a non-issue unless you want to unify `useRef` and `createRef`.

**Comment by ferdaber** [comment]
Since the issue itself was based on `useRef` and not `createRef`, I'm also okay with closing this as very few people out there directly modify the dereferenced pointer value.

**Comment by ferdaber** [comment]
I do wonder though if this overload pattern will pan out well, looking at the hooks documentation we see a few uses of `useRef` with a default value, in which case the value of `.current` may never be null but it can be kind of annoying for users to constantly use the non-null assertion operator to get past it.

**Comment by Kovensky** [comment]
@ferdaber `useRef` with a default value will always be mutable unless you explicitly both give it a generic argument and the generic argument does not include `| null`.

**Comment by ferdaber** [comment]
Ah, looking more closely shows that, and it's interesting that it's initialized as `undefined` in the React source without a parameter. Cool so everything's all peachy, it sounds like 👍

**Comment by bschlenk** [comment]
I think it is fine the way it is, just a bit odd that whether you include the `| null`, the type of `current` still can be null, just the mutability has changed. Also, the React docs always explicitly pass null, so is it even valid to omit the initialValue?

**Comment by Kovensky** [comment]
It's generally not, and at least in the past when I pointed that out in some piece of documentation the React team said that "even if it works" it's not intended to be ommitted.

It's why the first argument to `createContext` is required, for example, even if you do want a context to start with `undefined`. You have to actually pass `undefined`.

**Comment by ferdaber** [comment]
Looks like in some usages it's not using the parameter: https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables

**Comment by Kovensky** [comment]
@ferdaber that's because they're not caring about types there. What should the type of `current` in a `useRef()` (no arguments) be? `undefined`? `any`?

**Comment by Kovensky** [comment]
```ts
// you should always give an argument even if the docs sometimes miss them
// $ExpectError
const ref1 = useRef()
// this is a mutable ref but you can only assign `null` to it
const ref2 = useRef(null)
// this is also a mutable ref but you can only assign `undefined`
const ref3 = useRef(undefined)
// this is a mutable ref of number
const ref4 = useRef(0)
// this is a mutable ref of number | null
const ref4 = useRef<number | null>(null)
// this is a mutable ref with an object
const ref5 = useRef<React.CSSProperties>({})
// this is a mutable ref that can hold an HTMLElement
const ref6 = useRef<HTMLElement | null>(null)
// this is the only case where the ref is immutable
// you did not say in the generic argument you want to be able to write
// null into it, but you gave a null anyway.
// I am taking this as the sign that this ref is intended
// to be used as an element ref
const ref7 = useRef<HTMLElement>(null)
```

**Comment by ferdaber** [comment]
Yeah this DX works for me, and the documentation is A++ so I don't think most people will get confused on it. Thanks for this elaboration! Honestly you should just permalink this issue in the typedoc :)

**Comment by Kovensky** [comment]
There could be a case for supporting `useRef<T>()` and making it behave the same as `useRef<T | undefined>(undefined)`; but whatever ref you make with that still can't be used as an element ref, just as thread local storage.

**Comment by hielfx** [comment]
I'm getting this error with the following code:

~~~js
// ...
let intervalRef = useRef<NodeJS.Timeout>(null); // alos tried with const instead of let
// ... 
useEffect( () => {
    const interval = setInterval( () => { /* do something */}, 1000);
    intervalRef.current = interval; // In this line I'm getting the error
})
// ...
~~~

I'm new with both reack hooks and typescript (just trying them out together) so may code could be wrong

**Comment by ferdaber** [comment]
By default if you create a ref with a `null` default value and specify its generic parameter you're signaling your intent to have React "own" the reference. If you want to be able to mutate a ref that you own you should declare it like so:
```ts
const intervalRef= useRef<NodeJS.Timeout | null>(null) // <-- the generic type is NodeJS.Timeout | null
```

**Comment by hielfx** [comment]
@ferdaber Thanks for that!

**Comment by RoystonS** [comment]
Taking this a stage further and looking at the return type of `current`, should it perhaps be `T` rather than `T | null`?  With the advent of hooks, we _don't always_ have the case that all refs might be `null`, particularly in the frequent case where `useRef` is called with a non-null initializer.

Carrying on from the excellent example list in https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31065#issuecomment-446660394, if I write:

```
const numericRef = useRef<number>(42);
```
what should the type of `numericRef.current` be?  There's no _need_ for it to be `number | null`.

If we defined the types and functions as follows:

```
interface RefObject<T> {
  current: T;
}

function useRef<T>(initialValue: T): RefObject<T>;
function useRef<T>(initialValue: T|null): RefObject<T | null>;

function createRef<T>(): RefObject<T | null>;
```

that would produce the following usages and types:

```
const r1 = useRef(42);
// r1 is of type RefObject<number>
// r1.current is of type number (not number | null)

const r2 = useRef<number>(null);
// r2 is of type RefObject<number | null>
// r2.current is of type number | null

const r3 = useRef(null);
// r3 is of type RefObject<null>
// r3.current is of type null

const r4 = createRef<number>();
// r4 is of type RefObject<number | null>
// r4.current is of type number | null
```

Is anything there wrong?

For the answer to "what is the type of an unparameterised `useRef`?", the answer is that that call (despite the documentation) is incorrect, according to the React team.

**Comment by Jessidhia** [comment]
I added the overload with |null _specifically_ as a convenience overload for DOM/component refs, because they always start null, they are always reset to null on unmount, and you never reassign the current yourself, only React.

The readonly is there more to guard against logic errors than a representation of true JavaScript frozen object / getter-only property.

You can only make it by accident when you both give a generic argument that says you do not accept null while initializing the value to null anyway.

**Comment by RoystonS** [comment]
Ah, yes, I see now that `MutableRefObject<T>` already has the `| null` case removed, compared to `RefObject<T>`.  So the `useRef<number>(42)` case already works correctly. Thanks for the clarification!

What do we need to do with `createRef`?  Right now it returns an immutable `RefObject<T>`, which is causing problems in our codebase as we'd like to pass them around and use them in the same way as (mutable) `useRef` ref objects.  Is there a way we could tweak the `createRef` typing to allow it to form mutable ref objects?

**Comment by flq** [comment]
Don't tell me it's may and this is still broken? The whole point of useRef.current is to be able to set it.

**Comment by ferdaber** [comment]
We are not changing the type for `useRef` for the reasons detailed above (though `createRef` is still under discussion), did you have any questions on the rationale?

**Comment by flq** [comment]
Please excuse my comment, I didn't understand the full picture yet. If it's OK I'll delete it soon.

**Comment by ferdaber** [comment]
Yes, let's separate it out.

**Comment by nandorojo** [comment]
> If you want a mutable ref object that starts with a null value, make sure to also give `| null` to the generic argument. That will make it mutable, because you "own" it and not React.

Thanks for this! Adding null to the `useRef` generic solved it for me.

**Before**
``` javascript
const ref = useRef<SomeType>(null)
```
**After**
``` javascript
const ref = useRef<SomeType | null>(null)
```

**Comment by aappddeevv** [comment]
Did another comment about forwardRef components get created? Essentially, you can't forward a ref and change it's current value directly, which I think is part of the point of forwarding the ref.

**Comment by naorye** [comment]
I created the following hook:
```
export const useCombinedRefs = <T>(...refs: Ref<T>[]) =>
    useCallback(
        (element: T) =>
            refs.forEach(ref => {
                if (!ref) {
                    return;
                }

                if (typeof ref === 'function') {
                    ref(element);
                } else {
                    ref.current = element; // this line produces error
                }
            }),
        refs,
    );
```
And I get an error: "Cannot assign to 'current' because it is a read-only property."   
Is there a way to solve it without changing `current` to writeable?

**Comment by Jessidhia** [comment]
For that, you're going to have to cheat.

`(ref.current as React.MutableRefObject<T> ).current = element;`

Yes, this is a bit unsound, but it's the only case I can think of where this kind of assignment is deliberate and not an accident -- you're replicating an internal behavior of React and thus have to break the rules.

**Comment by guirenpei** [comment]
> We could modify `createRef` to have the same overload logic, but because there is no argument it'd have to be with a conditional type return. If you include `| null` it'd return `MutableRefObject`, if you don't it'd be (immutable) `RefObject`.

thanks a lot

**Comment by ArthurClemens** [comment]
> `(ref.current as React.MutableRefObject<T> ).current = element;`

should be:

`(ref as React.MutableRefObject<T> ).current = element;`

**Comment by s-h-a-d-o-w** [comment]
> you're replicating an internal behavior of React

Does that mean the [docs here](https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables) are outdated? Because they clearly describe this as an intended workflow, not internal behavior.

**Comment by ferdaber** [comment]
The docs in that case is referring to a ref that you control, but for refs that you pass as a `ref` attribute to an HTML element, then they should be readonly _to you_.

**Comment by s-h-a-d-o-w** [comment]
My bad, should've scrolled up further. Got an error about `readonly` for a ref I owned and assumed that that was the same case. (I use a different workflow now and can't reproduce it any more, unfortunately...)

Anyway, thank you very much for the explanation!

**Comment by sylvanaar** [comment]
I'm unclear if the `createRef` portion of the topic moved - but I will post here also.

I use a popular navigation library for React Native (React Navigation). In its documentation it commonly calls `createRef` and then mutates the ref. I am sure that this is because React isn't managing them (there is no DOM).

Should the type for React Native be different?

See: https://reactnavigation.org/docs/navigating-without-navigation-prop

**Comment by Imperyall** [comment]
@sylvanaar 
I also faced this problem when initializing navigation, have you found a solution?

**Comment by mexmirror** [comment]
Summarising what I just read: The only way to set a value to a ref created by `createRef<T>` is to cast it every time you use it? What's the reasoning behind that? In that case the `readonly` is practically preventing settings a value to the ref at all, therefore defeating the whole purpose of the function.

**Comment by jeffvandyke** [comment]
TLDR: If your initial value is `null` (details: or something else outside of the type parameter), then add `| null` to your type parameter, and that should make `.current` able to be assigned to like normal.

**Comment by hscheuerle** [comment]
More TLDR:
function useRef<T>(initialValue: T): MutableRefObject<T>;
function useRef<T>(initialValue: T|null): RefObject<T>;

**Comment by caleb-harrelson** [comment]
@sylvanaar / @Imperyall, I couldn't get typescript to be happy with `const isReadyRef = createRef<boolean | null>()`. I ended up changing the bit that sets the ref to true/false to cast it first.

    const mutableRef = isReadyRef as React.MutableRefObject<boolean>
    mutableRef.current = true

**Comment by Njaah-0** [comment]
I came to this issue because I got same error when using functions as a type. This was my code:
```
const pagesInitListener = useRef<() => void | null>(null);
pagesInitListener.current = () => { ... }
```
Solution was to wrap function into parenthesis, like this:
`const pagesInitListener = useRef<(() => void) | null>(null);`

**Comment by jdlk07** [comment]
I was trying to handle React Navigation's _Navigating without the navigation prop_  initialization as well, I ended up manually assigning the type like so since the type is surely going to be either a boolean or null.

```javascript
export const isReadyRef: React.MutableRefObject<boolean | null> = React.createRef()
```

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
